Packs/EWS/Integrations/SecurityAndCompliance/SecurityAndCompliance.ps1

```diff
@@ -382,423 +237,93 @@ function ParseSearchActionToEntryContext([psobject]$search_action, [int]$limit =
-
-        return $response_body
-
-        <#
-            .DESCRIPTION
-            Getting new access-token and refresh-token from Microsoft application based on the refresh-token we got from AccessTokenRequest() method.
-
-            .EXAMPLE
-            $client.RefreshTokenRequest()
-
-            .OUTPUTS
-            PSObject - Raw body response.
-
-            .LINK
-            https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-implicit-grant-flow#refreshing-tokens
-        #>
+        $this.app_id = $app_id
+        $this.organization = $organization
+        $this.upn = $upn
     }
 
-    [bool]IsDeviceCodeExpired(){
-        if (!$this.device_code){
-            return $true
+    CreateSession([string]$CommandName){
+        if ($null -eq $this.certificate) {
+            ReturnError "Error: For this command, a Certificate is required." | Out-Null
         }
-        $current_time = [int][double]::Parse((Get-Date -UFormat %s)) - 30
-        $valid_until = $this.device_code_creation_time + $this.access_token_expires_in
-
-        return $current_time -gt $valid_until
-
-        <#
-            .DESCRIPTION
-            Check if device-code expired.
-
-            .EXAMPLE
-            $client.IsDeviceCodeExpired()
-
-            .OUTPUTS
-            bool - True If device-code expired else False.
-
-            .LINK
-            https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-configurable-token-lifetimes#configurable-token-lifetime-properties-after-the-retirement
-        #>
-    }
-
-    [bool]IsAccessTokenExpired(){
-        if (!$this.access_token){
-            return $true
+        $cmd_params = @{
+            "AppID" = $this.app_id
+            "Organization" = $this.organization
+            "Certificate" = $this.certificate
+            "CommandName" = $CommandName
         }
-        $current_time = [int][double]::Parse((Get-Date -UFormat %s)) - 30
-        $valid_until = $this.access_token_creation_time + $this.access_token_expires_in
-
-        return $current_time -gt $valid_until
-        <#
-            .DESCRIPTION
-            Check if access-token expired.
-
-            .EXAMPLE
-            $client.IsAccessTokenExpired()
-
-            .OUTPUTS
-            bool - True If access-token expired else False.
-
-            .LINK
-            https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-configurable-token-lifetimes#configurable-token-lifetime-properties-after-the-retirement
-        #>
+        Connect-IPPSSession @cmd_params -WarningAction:SilentlyContinue | Out-Null
     }
 
-    RefreshTokenIfExpired(){
-        if ($this.access_token -and $this.IsAccessTokenExpired()) {
-            $this.RefreshTokenRequest()
+    CreateDelegatedSession([string]$CommandName){
+        if ($null -eq $this.delegated_password) {
+            ReturnError "Error: For this command, delegated access is required." | Out-Null
```

Comment:
Other commands can trip this error, but since it's handled in the session creation step, we shouldn't have other locations which would throw a similar error.